### PR TITLE
add PersistentVolumeClaim validation in ValidateStatefulSetSpec

### LINKED ISSUE #51306: StatefulSet validation should include pvc validation
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind bug

**What happened**:
the validation of statefulset is not complete, it not include the StatefulSetSpec.VolumeClaimTemplates validation, VolumeClaimTemplates  is an array of PersistentVolumeClaim, it should be checked.
**What you expected to happen**:
PersistentVolumeClaim in statefulset should be validated
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:
add validation for statefulset to cover all fields
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #51306

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by m1093782566** [comment]
ok-to-test

**Review by m1093782566 — COMMENTED** [review]
Can you add some UTs?

**Comment by hzxuzhonghu** [comment]
@m1093782566 sure

**Comment by m1093782566** [comment]
@hzxuzhonghu Can you check the UT failures?

**Comment by hzxuzhonghu** [comment]
@m1093782566 all tests paassed,  e2e test failure is a little random

File: pkg/apis/apps/validation/validation_test.go [hunk 1]
```diff
@@ -41,6 +50,28 @@ func TestValidateStatefulSet(t *testing.T) {
...
+		},
+		Resources: api.ResourceRequirements{
+			Requests: api.ResourceList{
+				api.ResourceName(api.ResourceStorage): resource.MustParse("10G"),
+			},
+		},
+		StorageClassName: &validClassName,
+	}
+
+	invalidClassName := "-invalid-"
```
Comment (m1093782566) [inline pkg/apis/apps/validation/validation_test.go]: nit: 

"-invalid-" -> "invalid"?

File: pkg/apis/apps/validation/validation.go [hunk 2]
```diff
@@ -39,6 +39,18 @@ func ValidateStatefulSetName(name string, prefix bool) []string {
 	return apivalidation.NameIsDNSSubdomain(name, prefix)
 }
 
+// Validates the given PersistentVolumeClaim of a statefulset.
+func ValidatePersistentVolumeClaimForStatefulset(pvc *api.PersistentVolumeClaim, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, msg := range apivalidation.ValidatePersistentVolumeName(pvc.ObjectMeta.GetName(), false) {
```
Comment (m1093782566) [inline pkg/apis/apps/validation/validation.go]: pvc.ObjectMeta.GetName() -> pvc.Name ?

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/apis/apps/validation/validation_test.go]: name should not consist of special character like "-" , so here construct one invalid name.

File: pkg/apis/apps/validation/validation.go [hunk 3]
```diff
@@ -39,6 +39,18 @@ func ValidateStatefulSetName(name string, prefix bool) []string {
...
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("metadata", "name"), pvc.ObjectMeta.GetName(), msg))
```
Comment (m1093782566) [inline pkg/apis/apps/validation/validation.go]: Same as above.

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (m1093782566) [reply pkg/apis/apps/validation/validation_test.go]: ACK.

File: pkg/apis/apps/validation/validation.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/apis/apps/validation/validation.go]: Yes, use  pvc.Name  instead.

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/apis/apps/validation/validation_test.go]: `names must be a DNS subdomain`

**Comment by m1093782566** [comment]
Some nits, otherwise LGTM.

**Comment by m1093782566** [comment]
/lgtm

ping @lavalamp for approval.

**Comment by hzxuzhonghu** [comment]
kindly ping @lavalamp for approval.

**Comment by errordeveloper** [comment]
@hzxuzhonghu would you mind to revise this and let's get someone to review it! Thanks.

**Comment by hzxuzhonghu** [comment]
ok

**Comment by hzxuzhonghu** [comment]
@errordeveloper rebased.
/assign @liggitt

**Comment by errordeveloper** [comment]
/test pull-kubernetes-e2e-kops-aws

@hzxuzhonghu I'm worried these failures could be related to your changes somehow...

**Comment by errordeveloper** [comment]
```
build/temp.linux-x86_64-2.7/check_libyaml.c:2:18: fatal error: yaml.h: No such file or directory
```
I think that had something to do with Bazel, please look into it.

**Comment by liggitt** [comment]
No, the kops failure is widespread across all PRs. The yaml compile error message is not related (the install falls back to not compile)

**Comment by hzxuzhonghu** [comment]
/retest

Does this meaningful? Any objections?

**Comment by liggitt** [comment]
in general, tightening validation is a breaking change (can break migration or upgrade if there are existing persisted objects that would fail the new validation)

**Comment by errordeveloper** [comment]
Should we have tests for that?

**Comment by liggitt** [comment]
> Should we have tests for that?

As we can, sure... the surface area of "every invalid value we didn't add validation for when the API was first released" tends to be a pretty ill-defined test target

**Pull request closed without merging**